### DISCUSSION TITLE
Portable / Standalone Executables

### DISCUSSION THREAD

**Discussion by denelon** [microsoft/winget-cli#1887] [Q&A]
I'm working on the specification for Portable / Standalone Executables #182

This discussion is to help consolidate discussion for this feature.
>Note: This is not the same as support for .zip or other compressed packages, but it is one of our prerequisites.

**Comment by denelon** [comment]
https://github.com/microsoft/winget-pkgs/pull/2594#issuecomment-1027015977

There has been an ask for the ability to rename a package.

I'm debating the value of using the "command" key in the manifest for renaming the package to the "suggested" command. We could also provide a "--rename" argument to allow the user to specify the name of their choice.

An example is GitLab Runner: According to their [documentation](https://docs.gitlab.com/runner/install/windows.html), they suggest renaming the executable after download.

**Reply by jedieaston** [reply]
This is probably a good idea, as there are many cases where developers name their downloads something not user friendly. At the very least, a symlink/shim to whatever is in the command field would be nice (this is [kind of what scoop does](https://scoop-docs.vercel.app/docs/concepts/The-Current-Version-Alias.html), and will give a cleaner user experience IMO).

**Reply by denelon** [reply]
This concept of "--rename" would only apply when installing a single portable / standalone program. If we leverage the "command" key in the manifest, we could use that to consistently rename the package. 

When a user is running `winget import`, they would essentially only get the default as specified in the manifest.

We've been discussing enhancements to the JSON schema for import / export. Some users have asked to be able to add other attributes like workloads for visual studio as "--override" works today. We wouldn't likely be able to export what was installed, but one can edit the packages.json file, and that would enable some of this kind of customization.

**Reply by jedieaston** [reply]
winget would have to keep track aware of what the package was renamed to so it could replace the correct files during `winget upgrade`, would it not? It seems like the --rename value could be filled into the packages.json since winget would hopefully know it anyway.

**Reply by denelon** [reply]
We could use the "same" mechanism from the manifest for upgrade if we use "command".

We would have to keep track in our local store of what was installed if the user leveraged the "--rename" argument.

**Reply by denelon** [reply]
If we were to use something akin to a symbolic link or shim, or App Execution Alias, that would be harder to maintain if the binary had a name change.

**Reply by denelon** [reply]
When we think about adding support for this in the Microsoft Store, it's best if we have the value to use for running the program in the manifest. I still see value in supporting "--rename" for developers who are using the Windows Package Manager CLI though 😊

**Reply by jedieaston** [reply]
Sounds good. In that case I'd hope that the command field was required for portable/standalone, as it's currently a nice-to-have.

**Reply by RobCannon** [reply]
I think installing a portable app with a versioned name and creating a symlink/shim for the friendly name should be considered a best practice.  It prevents issues with installing the new version if the old version is still in use, for instance.

**Reply by denelon** [reply]
I think the challenge with this is the risk of leaving files on the system as users upgrade. We would end up potentially with multiple entries in Apps & Features (one for each version of the package) and all of the files from the earlier versions. When we think about "upgrade" the idea is to simply upgrade the version. If the program is running, we want to be able to inform the user that it's running so they can shut it down / stop it before upgrading.

**Comment by majkinetor** [comment]
Maybe I am late to this party but here it is anyway...

Can we get some info on how this would work? Speaking as someone who created [hundreds](https://gist.github.com/choco-bot/a14b1e5bfaf70839b338eb1ab7f8226f) of [such packages](https://gist.github.com/majkinetor/a700c70b8847b29ebb1c918d47ee4eb1) on Chocolatey and the  [AU](https://github.com/majkinetor/au) framework (which I could make work with winget eventually) I have some questions and notes: 

1. Would I be able to pack random binaries into a package? This is is essentially what is needed for this to work, not only because of portability but also because this should be the preferred way of making packages given that they can then be cashed locally (via NAS, Artifactory, or Nexus for example) which makes them stable (not depending on network or vendor distribution model changes and having a stable list of previous versions) which mean that they can be **depended** on during build or CI/CD procedure. Without this option, you can't use winget for anything more serious but private OS customization.
2. Regarding leaving files on the system as users upgrade, that may be non-issue if upgraded files overwrite the previous one. Having it differently is funky, and makes almost 0 sense, as its not only about files but about migration configuration which can be hardcore or on some cases impossible if apps keep config in binary format like foobar2000.
3. It's way more important to have install than uninstall for portable packages. In many cases, it's not wise to remove those - for instance packages like influxdb or rundeck keep the production data in the install folder so by removing them you would destroy them, so in such cases it's best not to touch it.
4. What would be the size limit of such a package. Today, many, even simple tools can be 200+ MB in size.
5. Regarding keeping binaries on PATH, given that it is limited in size, one way to do it is to register them using [AppKey](https://github.com/chocolatey-community/chocolatey-packages/blob/master/extensions/chocolatey-core.extension/extensions/Register-Application.ps1) system. They will not be immediately available in CLI, but with simple [intervention](https://github.com/majkinetor/posh/blob/master/MM_Admin/Set-AppKeyAliases.ps1) they could (this could be supported by terminal natively). Other variant is to use shim to standard location which is added to PATH, but it has its own problems and surprises too, although it seems to mainly work.
6. Not sure what is the PowerShell support status, but this totally requires it, because some installation procedures may be [quite](https://github.com/majkinetor/au-packages/blob/master/rundeck/tools/helpers.ps1) [comprehensive](https://github.com/majkinetor/au-packages/blob/master/rundeck/tools/chocolateyInstall.ps1). 

I didn't yet use winget precisely because of this missing feature. If I can do more with this experience let me know and I can dive in deeper.

**Comment by denelon** [comment]
@majkinetor the portable [specification](https://github.com/microsoft/winget-cli/blob/master/doc/specs/%23182%20-%20Support%20for%20installation%20of%20portable%20standalone%20apps.md) covers a single executable binary.

1. I think you may be referring to the .zip [feature](https://github.com/microsoft/winget-cli/issues/140). Initially, we're thinking about a compressed file containing a single executable or a single installer and optionally other files. As we have additional clarity on other common scenarios, it can be extended.
2. I'd love to get your take on the upgrade mechanism we're considering. Using foobar2000 as an example might enlighten us, and justify some additional metadata in the manifest to help with upgrade scenarios. In general, we would like to ensure we have settings for default behaviors and pairs of opposing arguments so they can be overridden.
3. This sounds like more justification for a deeper dive on number two if our current thinking would be a bad user experience.
4. We haven't imposed size restrictions to date. Some installers are already in the GB range.
5. We've landed on using symbolic links in a single directory (per scope) on the path.
6. PowerShell will be supported as a native interface to the Windows Package Manager, but one of the primary design decisions has been to avoid scripting behavior. That does limit what the Windows Package Manager will "do" on its own. That doesn't mean one couldn't construct a package with additional steps necessary to achieve the desired outcome. The main difference (in my opinion) is the perception of liability and the risk associated with attempting to ensure all scripts are safe to run. 

The current [preview](https://github.com/microsoft/winget-cli/releases/tag/v1.3.1251-preview) as of today (May 10th, 2022) is our first release with experimental support for installing portable applications. It will require either using local manifests or some scripting to run manifests in PRs. If you're interested in testing it out, let us know. I'm sure some of the more prolific contributors would enjoy the collaboration. We also have an ongoing [chat](https://gitter.im/Microsoft/winget-pkgs) in Gitter.

**Comment by maphew** [comment]
Is there a feature or method to search package index for only portable apps and exes?

**Reply by maphew** [reply]
there is `winget search --tag portable` but I'm not sure how robust or reliable it is (there's only 28 items listed at the moment).

**Reply by denelon** [reply]
No, and "--tag" is just a metadata property that isn't correlated with "--installer-type". Feel free to create a feature request to be able to search by installer type.

**Reply by maphew** [reply]
I tried twice to open an issue and it seems to have failed. The 1st time I used the [...] at right and _reference in new issue_. While composing the description I used `ctrl-k` to insert a hyperlink, a dialog appeared on top with some unfamiliar and unrelated instructions. I hit `esc` and both the popup and the compose dialogs disappeared. I used reference in new issue again, and saved right away, then edited in the standard github new issue page. After finishing and submitting I returned to my Notifications page, and now there's a new item: https://github.com/microsoft/winget-cli/actions/runs/7747136718 whose title seems possiblty related (duplicate comment) but the content is very different(?)

![image](https://github.com/microsoft/winget-cli/assets/486200/42436a47-f170-42d9-8373-9b5f07070000)

_I know this post is better opened as new issue somewhere in GH as a UX bug, but I've no idea where that is, and need to write this down now while fresh in mind_

**Reply by denelon** [reply]
https://github.com/microsoft/winget-cli/issues/new/choose

From there you should be able to select "🚀 Feature Request / Idea"

**Comment by maphew** [comment]
> ... Feel free to create a feature request to be able to search by installer type.

ok, the new issue did work after all. However the system generated link which usually appears in context with the source discussion has been removed. Presumably as a consequence of the UX bug I described. Here's the request: https://github.com/microsoft/winget-cli/issues/4131

**Comment by maphew** [comment]
> _I know this post is better opened as new issue somewhere in GH as a UX bug, but I've no idea where that is, and need to write this down now while fresh in mind_

Found what I hope is the right spot and filed it: https://support.github.com/ticket/personal/0/2570211
